src/v2/guide/composition.md

```diff
@@ -0,0 +1,529 @@
+
+
+
+## Mixins
+
+### Mixins vs Extension
+
+**Mixins** work [like extension](#Merging-of-Instance-Properties), but allow you to define an array of configs instead of only one. For example, these two options do exactly the same thing:
+
+``` js
+extends: ComponentA
+```
+
+``` js
+mixins: [ComponentA]
+```
+
+So why have both then? Technically, you _could_ get everything done just with the `mixins` option, but we provide both to best describe their exact role in relation to the component:
+
+- **Extension is for building on a base component**, which will typically contain the template or render function. In the component we extend extending, we usually only add a few deviations from the default behavior.
+
+- **Mixins are for adding shared functionality to many components.** Unlike extended base components, mixins are typically not complete components by themselves. They do not contain a template or render function, leaving that to the components they're used in.
+
+### An Example Use Case
+
+Imagine we want to add a close button to multiple components. Upon click, this button would hide the component it's included in. We could create a mixin that provides a `<CloseButton/>` component wherever we want it to appear, like in this `AppNotification` component:
+
+``` js
+var AppNotification = {
+  mixins: [closeMixin],
+  template: `
+    <div class="notification">
+      <slot/>
+      <CloseButton/>
+    </div>
+  `
+}
+```
+
+And here's the mixin definition to make this work:
+
+``` js
+var closeMixin = {
+  // Register a close button for templates
+  components: {
+    CloseButton: {
+      template: `
+        <button v-on:click="$parent.isClosed = true">
+          X
+        </button>
+      `
+    }
+  },
+  // Track of whether the component is closed
+  data: function () {
+    return {
+      isClosed: false
+    }
+  },
+  // Simulate a v-if on the root element
+  beforeCreate: function () {
+    // Save the original, compiled render function
+    // (https://vuejs.org/v2/guide/render-function.html)
+    var origRender = this._render
+    // Replace the original render function
+    this._render = function _render () {
+      // If the element is closed...
+      return this.isClosed
+        // Render an empty text node
+        ? this._e()
+        // Otherwise call the original render
+        : origRender.apply(this, arguments)
+    }
+  }
+}
+```
+
+Looking at the mixin above, you might be thinking, "Wow, that looks complex!" You're right! Advanced component composition often requires deeper knowledge of both JavaScript and Vue's internals. To write the above mixin, you'd need to be familiar with:
+
+- JavaScript's [`Function.prototype.apply`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/apply).
+
+- Vue's internal `_render` method, which will either reference a user-provided render function or one created from a compiled, user-provided template.
```

Comment:
I'm in general against relying on any internal methods that start with an underscore - these methods may change behavior, be renamed or removed without notice (not connected to semver at all). Even with the warning tip below this is still in my opinion something that should not be demonstrated in the official docs.

Hacking the render function of the host component is also IMO prone to problems / conflicts if multiple mixins rely on similar strategies.

In this case it's more explicit to let the host component just add `<div v-if="!isClosed">`.

Reply:
I agree. When talking with other people I often see them already abusing some internals of Vue and whenever something breaks there, it's hard to figure out for them as they relied on the hack completely. 
I don't see it a good pattern to encourage people to go down that route.